### ISSUE TITLE
RFC: Improve error location

### ISSUE THREAD

**Issue by azu** [textlint/textlint#835]
Parts of #833

Currently, error message has only start position(`line` and `column` or `index`).
It prevents that underline actual error text.

Some editor plugin requires the full locations includes start and end.

Proposal

- Add `locator` function and `location` property
    - `locator` has `.range([number, number)` and `.location({ start, end })`
    - bike shading name

```js
const reporter = (context) => {
    const { Syntax, RuleError, fixer, locator, report, getSource } = context;
    return {
        [Syntax.Str](node) {
            const text = getSource(node);
            const matchRegexp = /\bfix\b/;
            if (!matchRegexp.test(text)) {
                return;
            }
            const index = text.search(matchRegexp);
            const length = "fix".length;
            report(
                node,
                new RuleError("Replaced", {
                    location: locator.range([index, index + length])
                })
            );
        }
    };
};
export default {
    linter: reporter,
    fixer: reporter
};
```

**Comment by azu** [comment]
We need to think about `location` naming.

ESLint's [context.report()](https://eslint.org/docs/developer-guide/working-with-rules#contextreport) use `loc` property.
In my experience, textlint prefer to use range than loc object.

**Comment by azu** [comment]
`locator` object should have following methods:

- `at(index)` == `range([index, index + 1])`
- `range([startIndex, endIndex])`
- `location({ start: { line, column }, end: { line, column })`

**Comment by azu** [comment]
Currently, textlint's `report` support 

- `index`
- `column`
- `line`

We can deprecate these options.

- `index` → `location: locator.at(index)`
- `column` → `location: location({ start: { line, column }, end: { line, column })`
- `line` → `location: location({ start: { line, column }, end: { line, column })`

However,  we will prserve these options for backward compatibility.
We will just add deprecation notice.

**Comment by azu** [comment]
Which is best `loc` or `location`.

`location` may conflict to `window.location`.
in a browser, Just type `location` that refer to `window.location`.

- [Location - Web API | MDN](https://developer.mozilla.org/ja/docs/Web/API/Location)

**Comment by azu** [comment]
We need to refactor `SourceLocation`.
It is very complex.

https://github.com/textlint/textlint/blob/fea2cf3e1f40042ec7c9aa2f1fc7d88eec11058c/packages/@textlint/kernel/src/core/source-location.ts#L29

**Comment by azu** [comment]
- [ ] need to change `isEqualMessage`
- [ ] 
https://github.com/textlint/textlint/blob/1234930f0f0056297e522959023cdb606f43fc5b/packages/@textlint/kernel/src/messages/filter-duplicated-process.ts#L9

**Comment by azu** [comment]
- [ ] Add `range` and `loc` to `TextlintMessage` interface

textlint result should have `range` and `loc` instead of `line`/`column`/`index`.
However, We will preverse `line`/`column`/`index` in the result.

https://github.com/textlint/textlint/issues/835#issuecomment-1000979549


https://github.com/textlint/textlint/blob/b2a03a1d1dc0049bf9c95d62ca5b9c9c61a8e2da/packages/%40textlint/types/src/Message/TextlintResult.ts#L7-L27

**Comment by azu** [comment]
Summary: error location patterns after this proposals

- Align to Node: `report(node, new RuleError(message))`
- range:  `report(node, new RuleError(message, { loc: locator.range([1,2]) } ))`
- location:  `report(node, new RuleError(message, { loc: locator.loc({ start: { line, column } , end: { line, column }) } ))`

**Comment by azu** [comment]
- [ ] refactor SourceLocation
- [ ] implement `loc` and `locator`
- [ ] fallback: convert `index`, `colimn`, `line` to `loc` info

**Comment by azu** [comment]
SourceLocation

- refactor adjust: separate line,column and fix
- create intermidiate represent  === locator result
- → to asbsolute with TxtNode

**Comment by azu** [comment]
https://github.com/textlint/textlint/pull/836#issuecomment-1001279968

![image](https://user-images.githubusercontent.com/19714/147426703-1e328675-a483-4b70-90ff-7b846c9bd84e.png)

filter ignored range behavior:

- A: Ignore
- B: Error
- C: Error
- D: Ignore
- E: Error

---

- Both(start and end) is contained: ignored
- Overlap: ignored
- Over point at least one(start or end): not ignored

**Comment by azu** [comment]
I've noticed that previous filtering logics has a bug.
https://github.com/textlint/textlint/issues/835#issuecomment-1001287841

Previously, only use start index to filter the matched error.
So, matched X was ignored unexpected.

matched X's start index is same with ignore's range end.

- Expected: shoud not ignore
- Actual: ignored 🐛 

![image](https://user-images.githubusercontent.com/19714/147446759-9f3ba0b3-ab58-42c5-a67b-2d47bc2a7c6a.png)

I will fix it in #836

**Comment by ThomasFKJorna** [comment]
This would be great!
